axiom/nr_distributed_trace.c

```diff
@@ -481,14 +481,23 @@ void nr_distributed_trace_set_app_id(nr_distributed_trace_t* dt,
 }
 
 void nr_distributed_trace_set_trace_id(nr_distributed_trace_t* dt,
-                                       const char* trace_id) {
+                                       const char* trace_id,
+                                       bool do_padding) {
   if (NULL == dt) {
     return;
   }
 
   nr_free(dt->trace_id);
   if (trace_id) {
-    dt->trace_id = nr_strdup(trace_id);
+    int len = nr_strlen(trace_id);
+    if (do_padding && len < NR_TRACE_ID_SIZE) {
+      int padding = NR_TRACE_ID_SIZE - len;
+      char* dest = (char*)nr_malloc(NR_TRACE_ID_SIZE + 1);
+      snprintf(dest, NR_TRACE_ID_SIZE+1, "%0*d%s", padding, 0, trace_id);
```

Comment:
For performance reasons `snprintf` should be replaced with manual padding. Manual byte manipulation is almost an order of magnitude more performant as shown by this simple program:
```
$ cat main.c
#include <stdio.h>
#include <string.h>
#include "nr_time.h"

#define NR_TRACE_ID_SIZE 32
#define N_ITER (1000 * 1000)

int main(int argc, char* argv[]) {
	int i;
	nrtime_t start, stop, sprintf_duration, strcpy_duration;
	char buf[NR_TRACE_ID_SIZE+1];
	char *cp;
	char *trace_id = "fedcba9876543210";

       	start = nr_get_time();
	for (i = 0; i < N_ITER; i++) {
		snprintf(buf, NR_TRACE_ID_SIZE+1, "%0*d%s", 16, 0, trace_id);
	}
       	stop = nr_get_time();
	sprintf_duration = stop - start;
	printf("%d x sprintf used " NR_TIME_FMT "us\n", i, sprintf_duration);

       	start = nr_get_time();
	for (i = 0; i < N_ITER; i++) {
		cp = buf;
		for (int p = 0; p < 16; p++) {
			*cp++ = '0';
		}
		strcpy(cp, trace_id);
	}
       	stop = nr_get_time();
	strcpy_duration = stop - start;
	printf("%d x manual used " NR_TIME_FMT "us\n", i, strcpy_duration);
	printf("sprintf vs manual diff " NR_TIME_FMT "us\n", sprintf_duration-strcpy_duration);
$ gcc main.c && ./a.out
1000000 x sprintf used 58901us
1000000 x manual used 10504us
sprintf vs manual diff 48397us
```

Reply:
changed in [e2c51c2](https://github.com/newrelic/newrelic-php-agent/pull/929/commits/e2c51c2595c8bd2620da4697790f957b9af4a4b4)